### PR TITLE
Speed up `Molecule` roundtrips

### PR DESCRIPTION
**Pull request by mattwthompson**
This is the first in a sequence of patches to improve performance when dealing with proteins or other large molecules/(bio-)macromolecules.

### CODE REVIEW

**Comment by mattwthompson** [comment]
I'm finding significant speedups in the re-creation of `Topology` and `Molecule` objects. This might not seem flashy or the bottleneck, but these operations crop up in numbers. Interchange immediately executes a deepcopy on the `Topology` it's passed, and much of the toolkit relies on the fast interconversion between the two objects (i.e. #699)

**Comment by mattwthompson** [comment]
```python3
In [1]: from openff.toolkit import Topology, Molecule

In [2]: t4 = Molecule.from_polymer_pdb("openff/toolkit/data/proteins/T4-protein.pdb").to_topology()

In [3]: %%timeit
   ...: Topology(t4)
   ...:
   ...:

# 91d03ae
187 ms ± 31.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# 0.12.0
1.26 s ± 44.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

**Comment by mattwthompson** [comment]
It'll be hard to squeeze out performance in force field loading since most of it it spent waiting for Pint to parse strings into `Quantity`s. Fortunately that issue should not scale with the complexity of the proteins being studied.

**Comment by mattwthompson** [comment]
I think this one's ready to go; the rest of this comment is a brief analysis. I used RDKit to generate proteins of different sizes from a naive sequence.

```python3
import string
import time

from rdkit import Chem

from openff.toolkit import Molecule, Topology

base_sequence = ""

for char in string.ascii_uppercase:
    try:
        Molecule.from_rdkit(Chem.MolFromSequence(char))
        base_sequence += char
    except:
        pass


for n in range(1, 51, 5):
    original_topology = Molecule.from_rdkit(
        Chem.MolFromSequence(n * base_sequence)
    ).to_topology()
    start = time.time()
    Topology(original_topology)
    diff = round(time.time() - start, 3)
    print(
        f"{n=}, {n * len(base_sequence)} residues, {original_topology.n_atoms} atoms, took {diff}"
    )
```

Running them and visualizing with a quick-and-dirty plot

```console
$git checkout upstream/main && python bench.py && git checkout protein-perf && python bench.py
HEAD is now at 5e4928b0 Fix "slow" gasteiger test (#1505)
n=1, 19 residues, 304 atoms, took 0.03
n=6, 114 residues, 1809 atoms, took 0.6
n=11, 209 residues, 3314 atoms, took 1.673
n=16, 304 residues, 4819 atoms, took 3.61
n=21, 399 residues, 6324 atoms, took 6.856
n=26, 494 residues, 7829 atoms, took 10.066
n=31, 589 residues, 9334 atoms, took 13.302
n=36, 684 residues, 10839 atoms, took 21.266
n=41, 779 residues, 12344 atoms, took 23.366
n=46, 874 residues, 13849 atoms, took 33.658
Previous HEAD position was 5e4928b0 Fix "slow" gasteiger test (#1505)
Switched to branch 'protein-perf'
n=1, 19 residues, 304 atoms, took 0.015
n=6, 114 residues, 1809 atoms, took 0.15
n=11, 209 residues, 3314 atoms, took 0.318
n=16, 304 residues, 4819 atoms, took 0.575
n=21, 399 residues, 6324 atoms, took 0.89
n=26, 494 residues, 7829 atoms, took 1.233
n=31, 589 residues, 9334 atoms, took 1.633
n=36, 684 residues, 10839 atoms, took 2.056
n=41, 779 residues, 12344 atoms, took 2.616
n=46, 874 residues, 13849 atoms, took 3.17
```

Overall, I think this is pretty good and not worth optimizing further. Doing anything interesting with a protein of a 1000 residues will surely be slow because of other steps in the pipeline.

image

This looks slightly superlinear, which is in line with our expectations. Plotting it on semi-log axes indicates this as well:

image

And plotting instead the time _per residue_ shows this more clearly:

image


The speedup probably asymptotes to a constant value with larger systems, but I didn't take the time to evaluate it. Most of the execution time is spend in or near RDKit and I'm not interesting in optimizing either here.

A main limitation of this benchmark is that the molecules had no conformers or partial charges - neither of which are trivial to generate with this approach, but each do contribute significantly to serialization performance as they're unit-bearing attributes. The other detail is that the scope is just `Topology(topology)`, which is important but only one step in the pipeline.

File: openff/toolkit/topology/molecule.py [hunk 1]
```diff
@@ -5037,15 +5019,14 @@ def add_atom(
         'C'
 
         """
-        atom_index = self._add_atom(
+        return self._add_atom(
```
Comment (j-wags) [inline openff/toolkit/topology/molecule.py:5022]: I think the original code here was a lot more readable - Could you revert it?

File: openff/toolkit/topology/molecule.py [hunk 2]
```diff
@@ -5084,15 +5065,14 @@ def add_bond(
         --------
         For an example of use, see :py:meth:`add_atom`.
         """
-        bond_index = self._add_bond(
+        return self._add_bond(
```
Comment (j-wags) [inline openff/toolkit/topology/molecule.py:5068]: I think the original code here was a lot more readable - Could you revert it?

**Review by j-wags — APPROVED** [review]
LGTM! This is a great speedup. Thanks!

**Pull request merged**